elasticsearch/src/main/scala/akka/stream/alpakka/elasticsearch/ElasticsearchFlowStage.scala

```diff
@@ -22,64 +22,118 @@ import akka.NotUsed
 import org.apache.http.message.BasicHeader
 import org.apache.http.util.EntityUtils
 
-object IncomingMessage {
+object IncomingIndexMessage {
   // Apply method to use when not using passThrough
-  def apply[T](id: Option[String], source: T): IncomingMessage[T, NotUsed] =
-    IncomingMessage(id, source, NotUsed)
+  def apply[T](source: T): IncomingMessage[T, NotUsed] =
+    IncomingMessage(id = None, source = Some(source), passThrough = NotUsed, operation = Index)
 
   // Apply method to use when not using passThrough
-  def apply[T](id: Option[String], source: T, version: Long): IncomingMessage[T, NotUsed] =
-    IncomingMessage(id, source, NotUsed, Option(version))
+  def apply[T](id: String, source: T): IncomingMessage[T, NotUsed] =
+    IncomingMessage(id = Some(id), source = Some(source), passThrough = NotUsed, operation = Index)
 
-  // Java-api - without passThrough
-  def create[T](id: String, source: T): IncomingMessage[T, NotUsed] =
-    IncomingMessage(Option(id), source)
+  // Apply method to use when using passThrough
+  def apply[T, C](source: T, passThrough: C): IncomingMessage[T, C] =
```

Comment:
Since we are already breaking source compatibility (`IncomingMessage` -> `IncomingIndexMessage`) we can remote this combinatorial explosion of factory methods. Lets keep only one and then have other combinations supported by calling `withXXX` builder methods.

Reply:
Specifying `source` and `passThrough` at the instance creation time is needed to fix type of `T` and `C`. If we don't specify them here, we need to call this method with type parameters like:
```scala
IncomingIndexMessage[Book, Offset](id).withSource(book).withPassTrough(offset)
```
instead of
```scala
IncomingIndexMessage(id, book, offset)
``` 
However Java API needs specify type parameters, anyway...

How do you think about it?

Reply:
The `withSource` and `withPassThrough` methods can change the type parameters of the `IncommingMessage`:

```scala
def withSource[S](source: S): IncomingMessage[S, C] =
  copy(source = Some(source))

def withPassThrough[P](passThrough: P): IncomingMessage[T, P] =
  copy(passThrough = passThrough)
```

With these it is possible to use builder methods without specifying the type parameters upfront:

```scala
IncomingIndexMessage("id")
  .withSource(Book())
  .withPassThrough("pass-through")
```

Reply:
That's right! I removed combinatorial factory methods except methods which take required parameters.